serde-generate/src/java.rs

```diff
@@ -10,737 +10,788 @@ use std::{
+                formats.len(),
+                Self::quote_types(formats, package_prefix)
+            ),
+            TupleArray { content, size } => format!(
+                "{} @com.facebook.serde.ArrayLen(length={}) []",
+                Self::quote_type(content, package_prefix),
+                size
+            ),
+            Variable(_) => panic!("unexpected value"),
+        }
     }
-    writeln!(out, "}}\n")
-}
 
-fn mangle_type(format: &Format) -> String {
-    use Format::*;
-    match format {
-        TypeName(x) => x.to_string(),
-        Unit => "unit".into(),
-        Bool => "bool".into(),
-        I8 => "i8".into(),
-        I16 => "i16".into(),
-        I32 => "i32".into(),
-        I64 => "i64".into(),
-        I128 => "i128".into(),
-        U8 => "u8".into(),
-        U16 => "u16".into(),
-        U32 => "u32".into(),
-        U64 => "u64".into(),
-        U128 => "u128".into(),
-        F32 => "f32".into(),
-        F64 => "f64".into(),
-        Char => "char".into(),
-        Str => "str".into(),
-        Bytes => "bytes".into(),
-
-        Option(format) => format!("option_{}", mangle_type(format)),
-        Seq(format) => format!("vector_{}", mangle_type(format)),
-        Map { key, value } => format!("map_{}_to_{}", mangle_type(key), mangle_type(value)),
-        Tuple(formats) => format!(
-            "tuple{}_{}",
-            formats.len(),
-            formats
-                .iter()
-                .map(mangle_type)
-                .collect::<Vec<_>>()
-                .join("_")
-        ),
-        TupleArray { content, size } => format!("array{}_{}_array", size, mangle_type(content)),
-        Variable(_) => panic!("unexpected value"),
+    fn quote_types(formats: &[Format], package_prefix: &str) -> String {
+        formats
+            .iter()
+            .map(|f| Self::quote_type(f, package_prefix))
+            .collect::<Vec<_>>()
+            .join(", ")
     }
-}
 
-fn needs_helper(format: &Format) -> bool {
-    use Format::*;
-    match format {
-        Option(_) | Seq(_) | Map { .. } | Tuple(_) | TupleArray { .. } => true,
-        _ => false,
+    fn output_trait_helpers(&mut self, registry: &Registry) -> Result<()> {
+        let mut subtypes = BTreeMap::new();
+        for format in registry.values() {
+            format
+                .visit(&mut |f| {
+                    if Self::needs_helper(f) {
+                        subtypes.insert(Self::mangle_type(f), f.clone());
+                    }
+                    Ok(())
+                })
+                .unwrap();
+        }
+        let prefix = if self.nested_class { "static " } else { "" };
+        writeln!(self.out, "{}final class TraitHelpers {{", prefix)?;
+        self.out.inc_level();
+        for (mangled_name, subtype) in &subtypes {
+            self.output_serialization_helper(mangled_name, subtype)?;
+            self.output_deserialization_helper(mangled_name, subtype)?;
+        }
+        self.out.dec_level();
```

Comment:
Also I just thought of a way to do scopes, although this code would need to be tweaked again for it to work.

Basically a function that lets you provide a closure:
```rust
    pub fn indented(&mut self, f: impl Fn(&mut IndentedWriter<T>) -> Result<()>) -> Result<()> {
        self.inc_level();
        let res = f(self);
        self.dec_level();
        res
    }
```

Reply:
Right, I think this would be the way to go, but after this PR, I'm not sure we need RAII. Mostly because there is no error management, but also coding pattern that seems to be emerging is to print `{` and call `inc_level`, then later to call `dec_level` and print `}`.